dimod/reference/samplers/exact_solver.py

```diff
@@ -177,28 +176,93 @@ def __init__(self):
         self.properties = {}
         self.parameters = {}
 
-    def sample_dqm(self, dqm, **kwargs):
+    def sample_dqm(self, dqm, **kwargs) -> SampleSet:
         """Sample from a discrete quadratic model.
 
         Args:
-            dqm (:obj:`~dimod.DiscreteQuadraticModel`):
+            dqm (:class:`~dimod.DiscreteQuadraticModel`):
                 Discrete quadratic model to be sampled from.
 
         Returns:
-            :obj:`~dimod.SampleSet`
+            :class:`~dimod.SampleSet`
 
         """
         Sampler.remove_unknown_kwargs(self, **kwargs)
         
-        n = dqm.num_variables()
-        if n == 0:
+        if not dqm.num_variables():
             return SampleSet.from_samples([], 'DISCRETE', energy=[])
 
-        possible_samples = as_samples((_all_cases_dqm(dqm), list(dqm.variables)))
-        energies = dqm.energies(possible_samples)
+        cases = _all_cases_dqm(dqm)
+        energies = dqm.energies(cases)
+        
+        return SampleSet.from_samples(cases, 'DISCRETE', energies)
+    
+    
+class ExactCQMSolver():
+    """A simple exact solver for testing and debugging code using your local CPU.
+
+    Notes:
+        This solver calculates the energy for every possible
+        combination of variable cases. It becomes slow very quickly
+        
+    Examples:
+        This example solves a CQM with 3 variables and 1 constraint.
+        
+        >>> from dimod import ConstrainedQuadraticModel, Binary
+        >>> cqm = ConstrainedQuadraticModel()
+        >>> x, y, z = Binary('x'), Binary('y'), Binary('z')
+        >>> cqm.set_objective(x*y + 2*y*z)
+        >>> cqm.add_constraint(x*y == 1)      # doctest: +SKIP
```

Comment:
is this solver not deterministic? I think if you use the `label=` keyword argument when adding the constraint you shouldn't need to do any of the `# doctest: +SKIP`